upt_macports/upt_macports.py

```diff
@@ -278,3 +278,32 @@ def package_versions(self, name):
             sys.exit(f'The command "{cmd}" failed. '
                      'Please make sure you have MacPorts installed '
                      'and/or your PATH is set-up correctly.')
+
+    def cpan_version(self, vers):
+        if vers[0] == "v":
+            start = 1
+        else:
+            start = 0
+
+        index = vers.find('.')
+        other_dot = vers[index+1:].find('.')
+        if index == -1 or other_dot != -1:
+            return vers[start:]
+        ret = vers[start:index]
+        index += 1
+        fractional = vers[index:]
+        index = 0
+        while index < len(fractional) or index < 6:
+            sub = fractional[index:index+3]
+            if len(sub) < 3:
+                sub += '0'*(3-len(sub))
+            ret = ret + '.' + str(int(sub))
+            index += 3
+        return ret
+
+    def needs_requirement(self, req, phase, frontend):
+        if frontend == 'cpan' and req.specifier:
+            print(req.specifier)
+            req.specifier = '>=' + self.cpan_version(req.specifier[2:])
```

Comment:
I don't think this can be generalized like this. Why do you assume the specifier always start with ```>=```? The way I read [the CPAN information on version ranges](https://metacpan.org/pod/CPAN::Meta::Spec#Version-Ranges), there are several options, including only a version number, or for example a one character operator like ```>```. 

I am not even sure if ```upt``` parses more complicated specifiers like ```>= 1.2, != 1.5, < 2.0``` but if so we would need to take that into account as well (perhaps also in the ```needs_requirement``` function, I don't know). So perhaps we need to use a regular expression here to separate the version from the operator.

Reply:
According to your link, CPAN has <, <=, ==, >=, > and !=. These are all supported by upt, which uses [PEP440](https://www.python.org/dev/peps/pep-0440/) internally.

If CPAN specifies "2.0", it means ">=2.0", and it's not properly parsed by upt-cpan, I'll need to fix this :)

Reply:
I understand that the operators are supported by ```upt```. My example specifier ```>= 1.2, != 1.5, < 2.0``` means that all version equal/above 1.2 and below 2 with the exception of 1.5 are supported. The question is whether the function ```needs_requirements``` in ```upt``` considers all these three constraints in the specifier when doing:

```
s = SpecifierSet(req.specifier)
compatible_versions = list(s.filter(versions))
```
, where versions is the list of versions that are packaged by the backend. I haven't checked that, but I was hoping one of you could save me the time trying ;)

Reply:
> If CPAN specifies "2.0", it means ">=2.0", and it's not properly parsed by upt-cpan, I'll need to fix this :)

Right, if CPAN specifies "2.0" that indeed means "=>2.0" but the conversion in the current implementation will do
```req.specifier = '>=' + self.cpan_version(req.specifier[2:])``` and result in ```req.specifier``` returned being ```>=0```.

Additionally, if the specifier contains multiple entries as in my example below; then we should do the version conversion on all of these specifiers.

Reply:
>> According to your link, CPAN has <, <=, ==, >=, > and !=. These are all supported by upt, which uses [PEP440](https://www.python.org/dev/peps/pep-0440/) internally.

okay, I just tried and indeed the more complicated specifiers do work correctly. I didn't look carefully enough at the ```filter``` function, so that's all good. The comment above still holds true though.

Reply:
Oh, I see your point, you make perfect sense. By the way, I think a similar issue exists in upt-cpan.

Would you know of a Perl package that has a "complex" specifier, such as the one from the documentation?

Reply:
@Steap a few examples are:
1. specifier with ```<```: Lexical-SealRequireHints under runtime/conflicts
2. with ```>=``` and ```!=```: Net-Twitter and Net-Twitter-Lite under runtime/requires
3. with ```==```: HTML-HeadParser-Liberal under runtime/requires

Reply:
Now this should work with all specifiers once it is fixed in upt-cpan

Reply:
@Korusuke that change looks good; you should have added tests for this though at the same time ;)